platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencySubstitutionRulesIntegrationTest.groovy

```diff
@@ -1612,4 +1658,64 @@ Required by:
             }
         }
     }
+
+    @Issue("https://github.com/gradle/gradle/issues/7594")
+    def "can substitute module with project and use withoutArtifactSelectors"() {
+        mavenRepo.module("com.external", "libB", "1.0")
+            .dependsOn("com.external", "libC", "1.0", "type")
+            .publish()
+
+        settingsFile << """
+            include("libC")
+        """
+
+        buildFile << """
+            plugins {
+                id("java-library")
+            }
+
+            ${mavenTestRepository()}
+
+            dependencies {
+                implementation("com.external:libB:1.0")
+            }
+
+            configurations.runtimeClasspath.resolutionStrategy.dependencySubstitution {
+                def sub = substitute(module("com.external:libC:1.0"))
+                    .using(project(":libC"))
+
+                if (${withoutArtifactSelectors}) {
+                    sub.withoutArtifactSelectors()
+                }
+            }
+
+            tasks.register("resolve") {
+                def files = configurations.runtimeClasspath.incoming.files
+                doLast {
+                    println(files*.name)
+                }
+            }
+        """
+
+        file("libC/build.gradle") << """
+            plugins {
+                id("java-library")
+            }
+        """
+
+        expect:
+        if (withoutArtifactSelectors) {
```

Comment:
I really don't like conditions in expectations, so I would do one of these...

1. Drop the fails path
2. Split the test into two separate tests into the failure/success cases (with and without `withoutArtifactSelectors`)
3. Check for failure, add `withoutArtifactSelectors` and check for success in one test
4. Split the test into two separate tests and make both pass in different ways. e.g., show that it's possible to select libC.type

Reply:
Done